### ISSUE TITLE
Pod stuck on Suspend state after recreation on Pod Group.

### ISSUE THREAD

**Issue by mbobrovskyi** [kubernetes-sigs/kueue#3683]
**What happened**:

Pod stuck on Suspend state after recreation on Pod Group.

```
2024-11-28T16:41:50.25358188Z	ERROR	jobframework/reconciler.go:429	Updating reclaimable pods	{"controller": "v1_pod", "namespace": "group/sts-e2e-pbznc", "name": "statefulset-sts-18b6e", "reconcileID": "d17d280a-8c61-4ec4-9d49-27b7398d8a74", "job": "group/sts-e2e-pbznc/statefulset-sts-18b6e", "gvk": "/v1, Kind=Pod", "error": "admission webhook \"vworkload.kb.io\" denied the request: status.reclaimablePods[5949e52e]: Required value: cannot be removed"}
sigs.k8s.io/kueue/pkg/controller/jobframework.(*JobReconciler).ReconcileGenericJob
	/workspace/pkg/controller/jobframework/reconciler.go:429
sigs.k8s.io/kueue/pkg/controller/jobs/pod.(*Reconciler).Reconcile
	/workspace/pkg/controller/jobs/pod/pod_controller.go:123
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:116
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:303
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:224
2024-11-28T16:41:50.253725505Z	ERROR	controller/controller.go:316	Reconciler error	{"controller": "v1_pod", "namespace": "group/sts-e2e-pbznc", "name": "statefulset-sts-18b6e", "reconcileID": "d17d280a-8c61-4ec4-9d49-27b7398d8a74", "error": "admission webhook \"vworkload.kb.io\" denied the request: status.reclaimablePods[5949e52e]: Required value: cannot be removed"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:224
```

**What you expected to happen**:
Pods running after recreation.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create pod group.

```
kubectl apply -f kueue-pod-group.yaml
```

```
---
apiVersion: v1
kind: Pod
metadata:
  name: sample-leader
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/pod-group-name: "sample-group"
  annotations:
    kueue.x-k8s.io/pod-group-total-count: "2"
spec:
  restartPolicy: Never
  containers:
  - name: sleep
    image: gcr.io/k8s-staging-perf-tests/sleep:v0.1.0
    args: ["30m"]
    resources:
      requests:
        cpu: "100m"
---
apiVersion: v1
kind: Pod
metadata:
  name: sample-worker
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/pod-group-name: "sample-group"
  annotations:
    kueue.x-k8s.io/pod-group-total-count: "2"
spec:
  restartPolicy: Never
  containers:
  - name: sleep
    image: gcr.io/k8s-staging-perf-tests/sleep:v0.1.0
    args: [ "30m"]
    resources:
      requests:
        cpu: "100m"

```

```
NAME            READY   STATUS    RESTARTS   AGE
sample-leader   1/1     Running   0          1s
sample-worker   1/1     Running   0          1s
```

2. Delete one of the pod.
```
kubectl delete pod sample-worker
```

3. Remove finalizer

```
NAME            READY   STATUS    RESTARTS   AGE
sample-leader   1/1     Running   0          24s
```

4. Create pod again.

```
kubectl apply -f kueue-pod-group.yaml
```

```
NAME            READY   STATUS            RESTARTS   AGE
sample-leader   1/1     Running           0          29s
sample-worker   0/1     SchedulingGated   0          3s
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by mbobrovskyi** [comment]
cc: @mimowo

**Comment by mimowo** [comment]
Actually, the fact that the replacement for a succeeded pods is suspended and gated is expected, for batch Jobs. We don't need to run a replacement if the previous task succeeded. 

I synced with @mbobrovskyi and this issue was originally motivated by the broken support for STS, but this is only a problem for "serving" workloads. For serving workloads we need to run the replacement pod even for succeeded pods. I created a dedicated issue https://github.com/kubernetes-sigs/kueue/issues/3690 to better reflect the problem.

Since the PodGroups were originally introduced for batch workloads we should not change the behavior and I proposed to introduce the "serving" annotation for PodGroups, implemented in https://github.com/kubernetes-sigs/kueue/pull/3686, so that we don't change the baseline behavior.

So, I believe we can close this issue and focus on https://github.com/kubernetes-sigs/kueue/issues/3690. I agree the error message could be better, but I think this is a separate issue. I think the  user-observable behavior is correct for batch workloads based on PodGroups.

cc @mbobrovskyi @tenzen-y 
/close
